src/gadget/is_zero.rs

```diff
@@ -0,0 +1,440 @@
+use halo2::{
+    circuit::{Chip, Region},
+    plonk::{Advice, Column, ConstraintSystem, Error, Expression, Selector, VirtualCells},
+    poly::Rotation,
+};
+use pasta_curves::arithmetic::FieldExt;
+use std::marker::PhantomData;
+
+pub(crate) trait IsZeroInstruction<F: FieldExt> {
+    /// Given a `value_diff` to be checked if it is zero:
+    ///   - witnesses `inv0(value_diff)`, where `inv0(X)` is 0 when `X` = 0,
+    ///     and `1/X` otherwise;
+    ///   - and returns nothing (result should be access by is_zero_expression)
+    fn is_zero(
+        &self,
+        region: &mut Region<'_, F>,
+        offset: usize,
+        value_diff: Option<F>,
+    ) -> Result<(), Error>;
+}
+
+#[derive(Clone, Debug)]
+pub(crate) struct IsZeroConfig<F> {
+    pub q_enable: Selector,
+    pub value_diff_inv: Column<Advice>,
+    /// This can be used directly for custom gate at the offset if `is_zero` is
+    /// called, it will be 1 if `value_diff` is zero, and 0 otherwise.
+    pub is_zero_expression: Expression<F>,
+}
+
+pub(crate) struct IsZeroChip<F> {
+    config: IsZeroConfig<F>,
+    _marker: PhantomData<F>,
+}
+
+impl<F: FieldExt> IsZeroChip<F> {
+    pub fn configure(
+        meta: &mut ConstraintSystem<F>,
+        q_enable: Selector,
+        value_diff: impl FnOnce(&mut VirtualCells<'_, F>) -> Expression<F>,
+        value_diff_inv: Column<Advice>,
+    ) -> IsZeroConfig<F> {
+        let mut is_zero_expression = None;
+
+        meta.create_gate(
+            "value_diff ⋅ (1 - value_diff ⋅ value_diff_inv) = 0",
```

Comment:
Consider adding this constraint `value_diff = 0 => valud_diff_inv = 0`
```suggestion
            "value_diff_inv ⋅ (1 - value_diff ⋅ value_diff_inv) = 0",
```

Reply:
Fixed in [ac331b5b](https://github.com/appliedzkp/zkevm-circuits/pull/12/commits/ac331b5b2689df4f22635e15c5caeb9d7d0c731e).